File: app/src/main/java/com/aaronbrecher/neverlate/Utils/DirectionsUtils.java
package com.aaronbrecher.neverlate.Utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;
import android.text.format.DateUtils;
import android.util.SparseArray;

import com.aaronbrecher.neverlate.Constants;
import com.aaronbrecher.neverlate.R;
import com.aaronbrecher.neverlate.models.Event;
import com.aaronbrecher.neverlate.models.EventLocationDetails;
import com.aaronbrecher.neverlate.models.retrofitmodels.EventDistanceDuration;
import com.aaronbrecher.neverlate.network.AppApiService;
import com.aaronbrecher.neverlate.network.AppApiUtils;
import com.google.android.gms.maps.model.LatLng;

import java.io.IOException;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;

/**
 * Class that contains functions to get distance information,
 * will use the Here api https://developer.here.com/
 */
public class DirectionsUtils {

    private SharedPreferences mSharedPreferences;
    private Location mLocation;
    private double mspeed;

    public DirectionsUtils(SharedPreferences preferences, Location location) {
        mSharedPreferences = preferences;
        mLocation = location;
        mspeed = Double.valueOf(mSharedPreferences.getString(Constants.SPEED_PREFS_KEY, "0.666667"));

    }

    /**
     * function to add distance information (Distance,Duration) to events
     * TODO for subscriptions - will execute this code always, if server response is that the user is not subscribed than will use as the crow flies data instead
     *
     * @param events   list of events from the calendar
     */
    public boolean addDistanceInfoToEventList(List<Event> events) {
        if (mLocation == null) return false;
        events = removeEventsWithoutLocation(events);
        SparseArray<List<Event>> transitTypes = splitEventListByTrasportType(events);
        return executeDrivingQuery(transitTypes.get(Constants.TRANSPORT_DRIVING))
                && executeTransitQuery(transitTypes.get(Constants.TRANSPORT_PUBLIC));
    }

    //Filter out all events that do not have a valid location
    private List<Event> removeEventsWithoutLocation(List<Event> eventList) {
        List<Event> filtered = new ArrayList<>();
        for (Event event : eventList) {
            if (event.getLocationLatlng() != null) {
                filtered.add(event);
            }
        }
        return filtered;
    }

    /**
     * This query will be to check for driving
     *
     * @return true if the data was added (even partially) false if not
     */
    private boolean executeDrivingQuery(List<Event> events) {
        return executeHereMatrixQuery(events, false);
    }

    /**
     * This query will be to check for public transportation data
     *
     * @return true if the data was added (even partially) false if not
     */
    private boolean executeTransitQuery(List<Event> events) {
        return executeHereMatrixQuery(events, true);
    }


    private boolean executeHereMatrixQuery(List<Event> events, boolean forPublicTransit) {
        if (events.size() == 0) return true;
        String origin = mLocation.getLatitude() + "," + mLocation.getLongitude();
        List<EventLocationDetails> destinations = convertEventListForQuery(events, forPublicTransit);
        AppApiService service = AppApiUtils.createService();

        Call<List<EventDistanceDuration>> request = forPublicTransit ? service.queryHerePublicTransit(origin, destinations)
                : service.queryHereMatrix(origin, destinations);
        try {
            Response<List<EventDistanceDuration>> response = request.execute();
            if (response.code() == 403) {
                addCrowFliesDistanceInfo(events);
                return true;
            }
            List<EventDistanceDuration> durationList = response.body();
            if (durationList == null || durationList.size() < 1) return false;
            for (int i = 0; i < durationList.size(); i++) {
                Event event = events.get(i);
                EventDistanceDuration distanceDuration = durationList.get(i);
                event.setDistance((long) distanceDuration.getDistance());
                event.setDrivingTime((long) distanceDuration.getDuration());
            }
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    private void addCrowFliesDistanceInfo(List<Event> events) {
        for (Event event : events) {
            Location eventLocation = new Location("never-late");
            eventLocation.setLatitude(event.getLocationLatlng().latitude);
            eventLocation.setLongitude(event.getLocationLatlng().longitude);
            long distance = (long) mLocation.distanceTo(eventLocation);
            if(distance <= 0) continue;
            long distanceInKilometers = distance/1000;
            long time = (long) (distanceInKilometers/mspeed) * 60;
            event.setDrivingTime(time);
            event.setDistance(distance);
        }
    }

    /**
     * convert the event list to a more concise representation to send to server, will include the latitude,
     * longitude, as well as the arrival time
     *
     * @param events             list of all the events
     * @param forPublicTransport if for a public transit request need to add the arrival time in iso-format
     * @return a list of the minimized event objects
     */
    private List<EventLocationDetails> convertEventListForQuery(List<Event> events, boolean forPublicTransport) {
        List<EventLocationDetails> destinations = new ArrayList<>();
        for (Event event : events) {
            long eventTime = GeofenceUtils.determineRelevantTime(event.getStartTime(), event.getEndTime());
            EventLocationDetails locationDetails = new EventLocationDetails(String.valueOf(event.getLocationLatlng().latitude),
                    String.valueOf(event.getLocationLatlng().longitude));
            if (forPublicTransport) {
                //TODO create iso-time for event location
                SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX");
                String formatted = dateFormat.format(new Date(eventTime));
                locationDetails.setArrivalTime(formatted);
            }
            destinations.add(locationDetails);
        }
        return destinations;
    }

    /**
     * Function to split the events into different driving catagories
     *
     * @return a map containing lists corresponding to all driving types
     */
    private SparseArray<List<Event>> splitEventListByTrasportType(List<Event> eventList) {
        List<Event> drivingEvents = new ArrayList<>();
        List<Event> publicEvents = new ArrayList<>();
        SparseArray<List<Event>> splitMap = new SparseArray<>();
        for (Event event : eventList) {
            switch (event.getTransportMode()) {
                case Constants.TRANSPORT_PUBLIC:
                    publicEvents.add(event);
                    break;
                case Constants.TRANSPORT_DRIVING:
                default:
                    drivingEvents.add(event);
                    break;
            }
        }
        splitMap.put(Constants.TRANSPORT_DRIVING, drivingEvents);
        splitMap.put(Constants.TRANSPORT_PUBLIC, publicEvents);
        return splitMap;
    }


    //TODO move these functions to a different class

    /**
     * Returns a readable string of distance to the event either in
     * Miles or KM
     */
    public static String getHumanReadableDistance(Context context, Long distance, SharedPreferences sharedPreferences) {
        //TODO add a shared prefs to miles or km and fix this accordingly
        boolean useMetric = true;

        String unitType = sharedPreferences.getString(context.getString(R.string.pref_units_key), "");
        if (context.getString(R.string.pref_units_imperial).equals(unitType)) {
            useMetric = false;
        }
        float km = distance.floatValue() / 1000;
        DecimalFormat df = new DecimalFormat("#.#");
        if (useMetric) {
            return df.format(km) + " " + context.getString(R.string.km_signature);
        } else {
            double miles = LocationUtils.kmToMiles(km);
            return df.format(miles) + " " + context.getString(R.string.miles_signature);
        }
    }

    /**
     * Returns a human readable representation of the time to leave to the
     * event
     *
     * @param timeTo    time until the event in seconds
     * @param eventTime time of the event in millis
     * @return Text of how much time to leave
     */
    public static String getTimeToLeaveHumanReadable(long timeTo, long eventTime) {
        timeTo = timeTo * 1000;
        long leaveTime = eventTime - timeTo;
        return DateUtils.getRelativeTimeSpanString(leaveTime, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS).toString();

    }


    public static String readableTravelTime(long travelTime) {
        int totalMinutes = (int) (travelTime / 60);
        int hours = totalMinutes / 60;
        int minutes = totalMinutes % 60;
        return hours + ":" + minutes;
    }
}

File: app/src/main/java/com/aaronbrecher/neverlate/Utils/DirectionsUtils.kt
package com.aaronbrecher.neverlate.Utils

import android.content.Context
import android.content.SharedPreferences
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.format.DateUtils
import android.util.SparseArray
import com.aaronbrecher.neverlate.AppExecutors

import com.aaronbrecher.neverlate.Constants
import com.aaronbrecher.neverlate.R
import com.aaronbrecher.neverlate.billing.BillingManager
import com.aaronbrecher.neverlate.billing.BillingUpdatesListener
import com.aaronbrecher.neverlate.interfaces.DistanceInfoAddedListener
import com.aaronbrecher.neverlate.models.Event
import com.aaronbrecher.neverlate.models.EventLocationDetails
import com.aaronbrecher.neverlate.models.HereApiBody
import com.aaronbrecher.neverlate.models.PurchaseData
import com.aaronbrecher.neverlate.network.AppApiService
import com.aaronbrecher.neverlate.network.*
import com.google.firebase.analytics.FirebaseAnalytics

import java.io.IOException
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date

/**
 * Class that contains functions to get distance information,
 * will use the Here api https://developer.here.com/
 */
class DirectionsUtils(mSharedPreferences: SharedPreferences,
                      private val mLocation: Location?,
                      private val distanceInfoAddedListener: DistanceInfoAddedListener,
                      private val context: Context) : BillingUpdatesListener {

    private val mspeed: Double = mSharedPreferences.getString(Constants.SPEED_PREFS_KEY, "0.666667")!!.toDouble()
    private val retrofitService: AppApiService = createRetrofitService()
    private lateinit var billingManager: BillingManager
    private lateinit var filteredEvents: List<Event>
    private val purchaseList: MutableList<PurchaseData> = ArrayList()
    private val appExecutors = AppExecutors()
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    /**
     * function to add distance information (Distance,Duration) to events
     *
     * @param events   list of events from the calendar
     */
    fun addDistanceInfoToEventList(events: List<Event>) {
        if (mLocation == null) return
        filteredEvents = removeEventsWithoutLocation(events)
//      TODO when want to start billing uncomment and change subscribe button to enabled in xml
//      billingManager = BillingManager(context, this)

//      TODO when want to start billing remove this
        appExecutors.networkIO().execute {
            val responseType = addDistanceFromHereApi()
            when (responseType) {
                QueryResponseType.SUCCESS -> distanceInfoAddedListener.distanceUpdated()
                QueryResponseType.FAILED,
                QueryResponseType.UNVERIFIED -> addCrowFliesDistanceInfo(filteredEvents)

            }
        }

    }

    override fun onBillingClientSetupFinished() {
        billingManager.getSubList(false)
    }

    override fun onPurchasesUpdated(purchases: List<PurchaseData>, wasAsync: Boolean) {
        appExecutors.networkIO().execute {
            purchaseList.addAll(purchases)
            val responseType = addDistanceFromHereApi()
            when (responseType) {
                QueryResponseType.SUCCESS -> distanceInfoAddedListener.distanceUpdated()
                QueryResponseType.FAILED -> addCrowFliesDistanceInfo(filteredEvents)
                QueryResponseType.UNVERIFIED -> {
                    if (wasAsync) addCrowFliesDistanceInfo(filteredEvents)
                    else mainThreadHandler.post { billingManager.getSubList(true) }
                }
            }
        }
    }

    override fun onBillingSetupFailed() {
        addCrowFliesDistanceInfo(filteredEvents)
    }

    private fun addDistanceFromHereApi(): QueryResponseType {
        val transitTypes = splitEventListByTransportType()
        val drivingQueryResponseType = executeDrivingQuery(transitTypes.get(Constants.TRANSPORT_DRIVING))
        val transitQueryResponseType = executeTransitQuery(transitTypes.get(Constants.TRANSPORT_PUBLIC))
        return if (drivingQueryResponseType == QueryResponseType.UNVERIFIED
                || transitQueryResponseType == QueryResponseType.UNVERIFIED) {
            QueryResponseType.UNVERIFIED
        } else if (drivingQueryResponseType == QueryResponseType.SUCCESS || transitQueryResponseType == QueryResponseType.SUCCESS) {
            QueryResponseType.SUCCESS
        } else {
            QueryResponseType.FAILED
        }
    }

    //Filter out all events that do not have a valid location
    private fun removeEventsWithoutLocation(eventList: List<Event>): List<Event> {
        return eventList.filter { it.locationLatlng != null }
    }

    /**
     * This query will be to check for driving
     * @return true if the data was added (even partially) false if not
     */
    private fun executeDrivingQuery(events: List<Event>): QueryResponseType {
        return executeHereMatrixQuery(events, false)
    }

    /**
     * This query will be to check for public transportation data
     * @return true if the data was added (even partially) false if not
     */
    private fun executeTransitQuery(events: List<Event>): QueryResponseType {
        return executeHereMatrixQuery(events, true)
    }


    private fun executeHereMatrixQuery(events: List<Event>, forPublicTransit: Boolean): QueryResponseType {
        if (events.isEmpty()) return QueryResponseType.SUCCESS
        val origin = mLocation!!.latitude.toString() + "," + mLocation.longitude
        val destinations = convertEventListForQuery(events, forPublicTransit)
        val request = if (forPublicTransit)
            retrofitService.queryHerePublicTransit(origin, HereApiBody(destinations, purchaseList))
        else
            retrofitService.queryHereMatrix(origin, HereApiBody(destinations, purchaseList))
        try {
            val response = request.execute()
            if (response.code() == 403) {
//                addCrowFliesDistanceInfo(events)
                return QueryResponseType.UNVERIFIED
            }
            val durationList = response.body()
            if (durationList == null || durationList.isEmpty()) return QueryResponseType.FAILED
            for (i in durationList.indices) {
                val event = events[i]
                val (distance, duration) = durationList[i]
                event.distance = distance.toLong()
                event.drivingTime = duration.toLong()
            }
        } catch (e: IOException) {
            e.printStackTrace()
            return QueryResponseType.FAILED
        }

        return QueryResponseType.SUCCESS
    }

    private fun addCrowFliesDistanceInfo(events: List<Event>) {
        events.forEach {
            val eventLocation = Location("never-late")
            eventLocation.latitude = it.locationLatlng!!.latitude
            eventLocation.longitude = it.locationLatlng!!.longitude
            val distance = mLocation!!.distanceTo(eventLocation).toLong()
            if (distance > 0) {
                val distanceInKilometers = distance / 1000
                val time = (distanceInKilometers / mspeed).toLong() * 60
                it.drivingTime = time
                it.distance = distance
            }
        }
        appExecutors.diskIO().execute { distanceInfoAddedListener.distanceUpdated() }
    }

    /**
     * convert the event list to a more concise representation to send to server, will include the latitude,
     * longitude, as well as the arrival time
     *
     * @param events             list of all the events
     * @param forPublicTransport if for a public transit request need to add the arrival time in iso-format
     * @return a list of the minimized event objects
     */
    private fun convertEventListForQuery(events: List<Event>, forPublicTransport: Boolean): List<EventLocationDetails> {
        val destinations = ArrayList<EventLocationDetails>()
        for (event in events) {
            val eventTime = GeofenceUtils.determineRelevantTime(event.startTime, event.endTime)
            val locationDetails = EventLocationDetails(event.locationLatlng!!.latitude.toString(),
                    event.locationLatlng!!.longitude.toString())
            if (forPublicTransport) {
                //TODO create iso-time for event location
                val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX")
                val formatted = dateFormat.format(Date(eventTime))
                locationDetails.arrivalTime = formatted
            }
            destinations.add(locationDetails)
        }
        return destinations
    }

    /**
     * Function to split the events into different driving catagories
     *
     * @return a map containing lists corresponding to all driving types
     */
    private fun splitEventListByTransportType(): SparseArray<List<Event>> {
        val drivingEvents = ArrayList<Event>()
        val publicEvents = ArrayList<Event>()
        val splitMap = SparseArray<List<Event>>()
        for (event in filteredEvents) {
            when (event.transportMode) {
                Constants.TRANSPORT_PUBLIC -> publicEvents.add(event)
                Constants.TRANSPORT_DRIVING -> drivingEvents.add(event)
                else -> drivingEvents.add(event)
            }
        }
        splitMap.put(Constants.TRANSPORT_DRIVING, drivingEvents)
        splitMap.put(Constants.TRANSPORT_PUBLIC, publicEvents)
        return splitMap
    }

    companion object {


        //TODO move these functions to a different class

        /**
         * Returns a readable string of distance to the event either in
         * Miles or KM
         */
        fun getHumanReadableDistance(context: Context, distance: Long, sharedPreferences: SharedPreferences): String {
            //TODO add a shared prefs to miles or km and fix this accordingly
            var useMetric = true

            val unitType = sharedPreferences.getString(context.getString(R.string.pref_units_key), "")
            if (context.getString(R.string.pref_units_imperial) == unitType) {
                useMetric = false
            }
            val km = distance.toFloat() / 1000
            val df = DecimalFormat("#.#")
            return if (useMetric) {
                df.format(km.toDouble()) + " " + context.getString(R.string.km_signature)
            } else {
                val miles = LocationUtils.kmToMiles(km)
                df.format(miles) + " " + context.getString(R.string.miles_signature)
            }
        }

        /**
         * Returns a human readable representation of the time to leave to the
         * event
         *
         * @param timeTo    time until the event in seconds
         * @param eventTime time of the event in millis
         * @return Text of how much time to leave
         */
        fun getTimeToLeaveHumanReadable(timeTo: Long, eventTime: Long): String {
            val timeToMillis = timeTo * 1000
            val leaveTime = eventTime - timeToMillis
            return DateUtils.getRelativeTimeSpanString(leaveTime, System.currentTimeMillis(), DateUtils.MINUTE_IN_MILLIS).toString()

        }


        fun readableTravelTime(travelTime: Long): String {
            val totalMinutes = (travelTime / 60).toInt()
            val hours = totalMinutes / 60
            val minutes = totalMinutes % 60
            return hours.toString() + ":" + minutes
        }
    }
}

enum class QueryResponseType {
    SUCCESS, FAILED, UNVERIFIED
}
